### DISCUSSION TITLE
which event to use for an autocmd to run a callback?

### DISCUSSION THREAD

**Discussion by andreacfromtheapp** [MunifTanjim/nui.nvim#326] [Q&A] [answered]
Hi @MunifTanjim 

ref: https://github.com/gacallea/chuck-nvim/issues/10
repo: https://github.com/gacallea/chuck-nvim

The following code runs fine if I invoke the function with a user command after vim is opened a chuck file but if I set autorun to true to invoke it at start of nvim (when opening a chuck file), the layout is created but none of the functions within the panes run.

```lua
-- this one creates the NUI layout and spawns the above
function M.chuck_ui(cmd, logfile)
  layout.chuck_layout:mount()
  layout.chuck_pane:on(events.BufEnter, function()
    start_chuck(cmd, logfile)
  end, { once = true })
  layout.shred_pane:on(events.BufEnter, function()
    shred_lines(logfile)
  end, { once = true })
  vim.cmd("wincmd w")

  layout.chuck_layout:update(layout.update_layout)
end
```

I have tried BufNew, BufAdd, VimEnter and more, to no avail. I'm out of ideas.

**Comment by andreacfromtheapp** [comment]
ps: if there is a way to run all code on layout:mount even better. the above doesn't seem right. it works but.

**Reply by MunifTanjim** [reply]
why not run the necessary functions before calling mount? 🤔

**Reply by andreacfromtheapp** [reply]
trying :)

**Reply by andreacfromtheapp** [reply]
running fn before doesn't seem to work :( 

just for completeness, here's the relevant code (sorry for the dump but it may help):

```lua
-- this calls the layout to set NUI elements in the UI
local function shred_nodes(line)
  local action = set_action(line)
  if line and action then
    layout.shreds_node(line, action)
    layout.shreds_tree:render()
  end
end

-- using tail to parse chuck's log file and trigger the above on a new line
local function shred_lines(logfile)
  local line = ""
  vim.fn.jobstart({ "tail", "-f", tostring(logfile) }, {
    on_stdout = function(_, data, _)
      data[1] = line .. data[1]
      line = data[#data]
      data[#data] = nil
      shred_nodes(data[#data])
    end,
  })
end

-- run chuck vm in a terminal split to see its output
local function start_chuck(cmd, logfile)
  local chuck_cmd = string.format("terminal %s 3>&1 2>&1 | tee %s", cmd, logfile)
  vim.cmd(chuck_cmd)
end

-- this one creates the NUI layout and spawns the above
function M.chuck_ui(cmd, logfile)
  layout.chuck_layout:mount()
  layout.chuck_pane:on(events.BufEnter, function()
    start_chuck(cmd, logfile)
  end, { once = true })
  layout.shred_pane:on(events.BufEnter, function()
    shred_lines(logfile)
  end, { once = true })
  vim.cmd("wincmd w")

  layout.chuck_layout:update(layout.update_layout)
end
```

**Reply by andreacfromtheapp** [reply]
ps: I may need to ue a conditional for when autorun is true or false. the above works great for false. need to find something for true.

**Reply by andreacfromtheapp** [reply]
ok I have found the issue:

if I manually click into the panes, the function run. using wincmd multiple times doesn't do the trick tho.

EDIT: finding the right event seems even more of a solution now.

**Reply by andreacfromtheapp** [reply]
or maybe run the functions on mount? I looked in the docs and it doesn't seem possible (afaiu)

**Reply by MunifTanjim** [reply]
I'm not entirely sure what you're trying to do.

Can you explain a bit (without codes or technical terms)? For example: I wanna run a command that will open two splits and do something and show something on the top split and do something to the bottom split.

**Reply by andreacfromtheapp** [reply]
I'll try to rephrase the opening post:

The plugin needs both chuck (chuck_pane) and shreds list (shred_pane) to run in order to work (since it is basically a _pseudo-IDE_).

Now:

- the user can launch the UI manually (`autorun = false`) with `ChuckLoop` command. All good here.
- the user can have it launch the UI at nvim startup when opening a chuck file (`autorun = true`) (with the autocmd invoking `ChuckLoop` on user's behalf).

So far:

- manualy works as intended with `BufEnter`.
- autorun is the problematic one. It does launch the UI (so `ChuckLoop` is run fine) **BUT** it needs a user to click the split buffers to run the callbacks. **Not ideal**.

My issue:

- I'd need to find a way to make `autorun = true` work without the user needing to click anything when opening a chuck file at nvim startup.

Possible solution (just my first guess):

- what is the right event to add here for the `autorun = true`?

Please and thank you

**Reply by MunifTanjim** [reply]
> So far:
> 
> - manualy works as intended with `BufEnter`.
> - autorun is the problematic one. It does launch the UI (so `ChuckLoop` is run fine) **BUT** it needs a user to click the split buffers to run the callbacks. **Not ideal**.
> 
> My issue:
> 
> - I'd need to find a way to make `autorun = true` work without the user needing to click anything when opening a chuck file at nvim startup.
> 
> Possible solution (just my first guess):
> 
> - what is the right event to add here for the `autorun = true`?

This part actually describes your solution for a problem I don't yet understand. This leads to the XY Problem: https://xyproblem.info/

I'm trying to understand the actual problem so that I can suggest a possibly better solution. Why do you need to call some codes on `BufEnter`? What do those codes do?

**Reply by andreacfromtheapp** [reply]
The code I [shared earlier](https://github.com/MunifTanjim/nui.nvim/discussions/326#discussioncomment-8206887) is commented and I thought it was clear. My bad.

I'll try recoding a video after I finish lunch. They say a picture is worth a thousands words.

**Reply by andreacfromtheapp** [reply]
Hi @MunifTanjim :)

I managed to do what I needed! The `:on(event, yada)` stuff was an attempt because I wasn't sure this was possible with `:mount()`, but I thought I'd give it a go before wrinting more. This does what I need. The more I learn about NUI the better it gets.

```lua
-- this one creates the NUI layout and spawns the above
function M.chuck_ui(cmd, logfile)
  layout.chuck_layout:mount()
  layout.chuck_pane:mount(start_chuck(cmd, logfile), { once = true })
  layout.shred_pane:mount(shred_lines(logfile), { once = true })
  vim.cmd("wincmd w")
end
```

Except for one thing. Now the issue is inverted. Launching the UI **manually** (cmd: ChuckLoop) opens the UI and runs the commands just fine **but** hides the layout right away. If I run the command a second time, it shows up.

If this is also unclear, I'll try to explain better in a follow up. Sorry about my bad communication skills.

Thank you for your help so far.

**Reply by MunifTanjim** [reply]
`:mount` doesn't take any arguments.

These

```lua
  layout.chuck_pane:mount(start_chuck(cmd, logfile), { once = true })
  layout.shred_pane:mount(shred_lines(logfile), { once = true })
```

are actually same as

```lua
  start_chuck(cmd, logfile)
  layout.chuck_pane:mount()
  shred_lines(logfile)
  layout.shred_pane:mount()
```

ref: https://github.com/MunifTanjim/nui.nvim/tree/main/lua/nui/popup#popupmount

**Reply by MunifTanjim** [reply]
And `layout:mount()` automatically calls `:mount` for all the children.

So,

```lua
  layout.chuck_layout:mount()
  start_chuck(cmd, logfile)
  layout.chuck_pane:mount()
  shred_lines(logfile)
  layout.shred_pane:mount()
```

is same as:

```lua
  layout.chuck_layout:mount()
  start_chuck(cmd, logfile)
  shred_lines(logfile)
```

**Reply by andreacfromtheapp** [reply]
wtf?! lucky accident then (at least I had understood the docs the first time around when I thought it was not possible). precious info nonetheless as it may help me understand why it hides on manual command and not with the autocmd.

**Reply by MunifTanjim** [reply]
What are you expecting `start_chuck` to do? I don't think it's doing what you're expecting it to do.

Running `:terminal <cmd>` on a split window replaces the buffer of that window. And that would break the `NuiSplit`, because it tracks the buffer in `.bufnr` property and it goes out of sync when you replace the buffer on that window by running `:terminal <cmd>`. So `split.bufnr` holds a differnt buffer number than the buffer that is actually being shown in that window. So `NuiSplit` won't behave as expected anymore after that.

**Reply by MunifTanjim** [reply]
- https://github.com/gacallea/chuck-nvim/blob/cd3ca820913a2cb46866eb4ee79cea0d765a84c6/lua/chuck-nvim/ui/layout.lua#L23-L55
- https://github.com/gacallea/chuck-nvim/blob/cd3ca820913a2cb46866eb4ee79cea0d765a84c6/lua/chuck-nvim/ui/layout.lua#L82-L94

These things are not meant to be created once and stored in the module. Because once you close the layout/splits those instances of `NuiSplit`/`NuiLayout` becomes unusable.

They should be created inside a function, used once and then discarded.

_P.S. They can be re-used, but that needs careful handling... You need to dig deeper into how buffers and windows behave in Neovim._

For example, check: https://github.com/jackMort/pommodoro-clock.nvim/blob/eb7a59109c82917ba1c386f730ecdb9ef778b8a5/lua/pommodoro-clock.lua#L195-L216

A new `NuiPopup` is created inside the `show_popup` function. Whenever user calls `show_popup`, a new `NuiPopup` is created.

**Reply by andreacfromtheapp** [reply]
I see. I may have picked the wrong tool? (nothing against NUI, I may have misunderstood its nature).

Basically the idea is to use "fixed splits" to represent data like this ui does:

<img width="953" alt="Screenshot 2024-01-22 at 3 28 18 PM" src="https://github.com/MunifTanjim/nui.nvim/assets/3269984/8b71f276-6812-4255-937e-7b278008809a">

**Reply by andreacfromtheapp** [reply]
ps: before this whole Discussion topic, it worked as intended on manual. I may go back to that and try to find an event that fits teh autocmd. Just speaking my mind out here. Nothing against you or NUI.

**Reply by andreacfromtheapp** [reply]
> These things are not meant to be created once and stored in the module. Because once you close the layout/splits those instances of NuiSplit/NuiLayout becomes unusable.

this is not an issue, because the UI is created on demand and destroyed on exit only. see above screenie for "usage".

**Reply by MunifTanjim** [reply]
> Basically the idea is to use "fixed splits" to represent data like this ui does:

You can absolutely do that with or without `nui.nvim`.

> I see. I may have picked the wrong tool? (nothing against NUI, I may have misunderstood its nature).

`nui.nvim` is nothing but a convenient wrapper around Neovim's APIs. If you use it or not, you'll have to deal with the problems anyway.

When you open a split, nobody is gonna stop the user from going into that split and running `:quit`. How you handle that scenario is up to you.

> this is not an issue, because the UI is created on demand and destroyed on exit only.

What happens when I quit the splits and run `:ChuckLoop` again?

---

Anyway, I really can't help you unless I have a minimal script that I can run to reproduce the bug. I can't manage the time to debug a whole new plugin...

If `nui.nvim` has any bug or if you have question about how to use `nui.nvim`, I'm happy to help. But I don't think your problems are related to `nui.nvim`, those are just general Neovim related problems.

**Reply by andreacfromtheapp** [reply]
All fair points. I should not burden you with my own learning. I may have overdone this whole plugin. Perhaps running those things in tmux is a better idea. I want to thank you for your time, patience, and support. I'll get back to this tomorrow with a fresh mind. Cause I may need to change a lot.

That said, I don't think it's possible to provide a minimal script without the chuck dependency. At least I wouldn't know what to provide at this stage of both my skills and plugin state. Perhaps using splits isn't ideal here. I tried "pure nvim" at first (split, vsplit and so on) but didn't like some things and looked for a tui and found nui.

Again, thank you so very much <3

**Reply by MunifTanjim** [reply, marked as answer]
Just as a closing note, I'm not entirely sure how `chuck` cli command works... but you should try to separate it from the UI concerns. UI (splits) can be closed and opened again and again. But I'd assume you don't want to run `chunk` cli command multiple times. Also in `shred_lines`, I assume you're trying to listen to changes in the log file... you shouldn't do that every time UI is opened, you only need to do that once.

Once you separate out those two concerns, a lot of the problems you're facing will disappear.

And about `:terminal <cmd>` it creates a new terminal buffer, you can store that terminal buffer number somewhere and then just change the buffer of the split to that terminal buffer when you re-open the UI.

**Reply by andreacfromtheapp** [reply]
@MunifTanjim yes!!! https://github.com/gacallea/chuck-nvim/commit/68811a3fe9e8ce795b368faccce9dacc1db2092f -- will make a release and asciinema to add to the README so you can see it. One last thing I need to fix is that split doesn't "scroll" (to show last lines)

**Reply by andreacfromtheapp** [reply]
ps: not the best separation of concerns but at least it's something. I wanted to make it work first, and I'm now looking into turning those jobs to hidden buffers I can attach to.
